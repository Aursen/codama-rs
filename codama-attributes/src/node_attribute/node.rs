use super::NodeAttributeParse;
use codama_nodes::*;
use codama_syn_helpers::syn_traits::*;

impl NodeAttributeParse for Node {
    fn from_meta(path: &syn::Path, meta: &syn::meta::ParseNestedMeta) -> syn::Result<Node> {
        match path.last_str().as_str() {
            // Type nodes.
            "boolean_type" => BooleanTypeNode::from_meta(path, meta),
            "number_type" => NumberTypeNode::from_meta(path, meta),
            "public_key_type" => PublicKeyTypeNode::from_meta(path, meta),
            _ => return Err(meta.error("unrecognized node")),
        }
    }
}
