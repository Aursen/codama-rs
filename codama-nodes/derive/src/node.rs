use crate::lowercase_first_letter;
use codama_errors::CodamaResult;
use codama_syn_helpers::extensions::*;
use proc_macro2::TokenStream;
use quote::quote;

pub fn expand_attribute_node(input: &syn::DeriveInput) -> CodamaResult<TokenStream> {
    input.as_struct()?;
    let item_name = &input.ident;
    let kind = lowercase_first_letter(&item_name.to_string());

    Ok(quote! {
        #[derive(codama_nodes_derive::Node, core::fmt::Debug, core::cmp::PartialEq, core::clone::Clone, serde::Serialize, serde::Deserialize)]
        #[serde(tag = "kind", rename = #kind)]
        #[serde(rename_all = "camelCase")]
        #input
    })
}

pub fn expand_derive_node(input: &syn::DeriveInput) -> CodamaResult<TokenStream> {
    input.as_struct()?;
    let item_name = &input.ident;
    let (pre_generics, post_generics) = &input.generics.block_wrappers();
    let kind = lowercase_first_letter(&item_name.to_string());

    Ok(quote! {
        impl #pre_generics crate::NodeTrait for #item_name #post_generics{
            const KIND: &'static str = #kind;
        }
    })
}
