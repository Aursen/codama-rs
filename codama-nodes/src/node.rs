use crate::{
    AccountNode, CountNode, DefinedTypeNode, DiscriminatorNode, ErrorNode, InstructionAccountNode,
    InstructionArgumentNode, InstructionByteDeltaNode, InstructionNode,
    InstructionRemainingAccountsNode, LinkNode, PdaNode, PdaSeedNode, ProgramNode,
    RegisteredContextualValueNode, RegisteredTypeNode, RegisteredValueNode, RootNode,
};
use codama_nodes_derive::IntoEnum;
use serde::{Deserialize, Serialize};

#[derive(IntoEnum, Debug, PartialEq, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum Node {
    // Node unions.
    ContextualValue(RegisteredContextualValueNode),
    Count(CountNode),
    Discriminator(DiscriminatorNode),
    Link(LinkNode),
    PdaSeed(PdaSeedNode),
    Type(RegisteredTypeNode),
    Value(RegisteredValueNode),

    // Nodes.
    Account(AccountNode),
    DefinedType(DefinedTypeNode),
    Error(ErrorNode),
    Instruction(InstructionNode),
    InstructionAccount(InstructionAccountNode),
    InstructionArgument(InstructionArgumentNode),
    InstructionByteDelta(InstructionByteDeltaNode),
    InstructionRemainingAccounts(InstructionRemainingAccountsNode),
    Pda(PdaNode),
    Program(ProgramNode),
    Root(RootNode),
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{NumberTypeNode, U32};

    #[test]
    fn type_node_to_json() {
        let node: Node = RegisteredTypeNode::Number(NumberTypeNode::le(U32)).into();
        let json = serde_json::to_string(&node).unwrap();
        assert_eq!(
            json,
            r#"{"kind":"numberTypeNode","format":"u32","endian":"le"}"#
        );
    }

    #[test]
    fn type_node_from_json() {
        let json = r#"{"kind":"numberTypeNode","format":"u32","endian":"le"}"#;
        let node: Node = serde_json::from_str(json).unwrap();
        assert_eq!(
            node,
            Node::Type(RegisteredTypeNode::Number(NumberTypeNode::le(U32)))
        );
    }

    #[test]
    fn defined_type_to_json() {
        let node: Node = DefinedTypeNode::new("myType", NumberTypeNode::le(U32)).into();
        let json = serde_json::to_string(&node).unwrap();
        assert_eq!(
            json,
            r#"{"kind":"definedTypeNode","name":"myType","type":{"kind":"numberTypeNode","format":"u32","endian":"le"}}"#
        );
    }

    #[test]
    fn defined_type_from_json() {
        let json = r#"{"kind":"definedTypeNode","name":"myType","type":{"kind":"numberTypeNode","format":"u32","endian":"le"}}"#;
        let node: Node = serde_json::from_str(json).unwrap();
        assert_eq!(
            node,
            Node::DefinedType(DefinedTypeNode::new("myType", NumberTypeNode::le(U32)))
        );
    }
}
