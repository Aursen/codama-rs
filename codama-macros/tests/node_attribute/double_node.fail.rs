use codama_macros::node;

#[node(number_type(u32, le), public_key_type())]
pub struct Foo(usize);

fn main() {}
