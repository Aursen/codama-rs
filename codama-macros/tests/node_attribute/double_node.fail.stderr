error: node is already set
 --> tests/node_attribute/double_node.fail.rs:3:30
  |
3 | #[node(number_type(u32, le), public_key_type())]
  |                              ^^^^^^^^^^^^^^^^^
